# Assignment 9 

# Python Function Assignment -1

**Q.1. What is a lambda function in Python, and how does it differ from a regular function?**

**Answer:** A function without a name is called **'anonymous function'**. So far, the functions we wrote were defined using the keyword 'def'. But anonymous function are not defined using 'def'. They are defined using the keyword **lambda** and hence they are also called **'Lambda Functions'**. Let's take a normal function that returns square value.

In [14]:
def square_def(x):
    return x*x

The same function can be written as anonymous function as:

In [ ]:
lambda x: x*x

Observe the keyword 'lambda'. This represents that an anonymous function is being created. After that, we have written an argument of function, i.e. 'x'. Then colon (:) represents the beginning of the function that contains an expression x*x. Please observe that we did not use any name for the function here. So, the format of lambda functions is:

In [ ]:
lambda argument_list: expression

Normally, if a function returns some value, we assign that value to a variable as:

In [ ]:
y = square(5)

but, lambda functions return a function and hence they should be assigned to a function as:

In [ ]:
f = lambda x: x*x

Here, 'f' is the function name to which the lambda expression is assigned. Now, if we call the function f() as:

In [ ]:
value = f(5)

Now, 'value' contains the square value of, i.e. 25.

In [6]:
# a lambda function to calculate square value
f = lambda x: x*x # write lambda function
value = f(5) # call lambda function
print('Square of 5 = ', value) # display reult

Square of 5 =  25


|def defined functions|lambda functions|
|---------------------|----------------|
|1.Easy to interpret|1.nterpretation might be tricky|
|2.Can consists of any number of execution statements inside the function definition.|2.The limited operation can be performed using lambda functions.|
|3.To return an object from the function, return should be explicitly defined.|3.No need of using the return statement.|
|4.Execution time is relatively slower for the same operation performed using lambda functions.|4.Execution time of the program is fast for the same operation.|
|5.Defined using the keyword def and holds a function name in the local namespace.|5.Defined using the keyword lambda and does not compulsorily hold a function name in the local namespace.|

**Q.2.Can a lambda function in Python have multiple arguments? If yes, how can you define and use
them?**

**Answer:** Yes, lambda function have multiple arguments. Python lambda can be used with multiple arguments and these arguments are used in evaluating an expression to return a single value. A Python lambda function is used to execute an anonymous function, an anonymous meaning function without a name.

Here, we can create a lambda function with multiple arguments in Python, let’s see an example with two arguments, add them and return the result. Here, I have created a add variable and assigned a lambda function to it.

In [8]:
# Lambda example with two arguments
add = lambda x, y : x + y
print(add(10, 20)) 

30


Alternatively, you can also write the statement as follows. This is called inline execution. For inline invocation, we surround the lambda function with parentheses and place the values for the arguments next to it enclosed within parentheses.

In [9]:
result = (lambda x, y : x + y)(10,20)
print(result)

30


Let’s create a lambda function with **multiple arguments**, and take the values for these arguments from the list, To pass values to these arguments you need to use two iterables.

In [10]:
# Create two lists with numbers 
numbers1 = [2, 4, 5, 6, 3]
numbers2 = [1, 3, 2, 2, 4]

# Create lambda function that takes two arguments
add_fun = lambda x, y: x+y

# Use lambda with map() function
add_result = list(map(add_fun, numbers1, numbers2))
print(add_result) 

[3, 7, 7, 8, 7]


Here, the python lambda takes multiple arguments x, y and adds their values. The map() function applies this lambda function to each item of the numbers1 & numbers2 lists, and the result is a new iterator containing the sum of both numbers element-wise.

In [11]:
# Create string list
myStrings1 = ["A","B","C","D"]
myStrings2 = ["a","b","c","d"]
print("myStrings1:",myStrings1)
print("myStrings2:",myStrings2)

# Use lower() function
lower_result = list(map(lambda x,y: x + y, myStrings1, myStrings2))
print("Result:",lower_result)

myStrings1: ['A', 'B', 'C', 'D']
myStrings2: ['a', 'b', 'c', 'd']
Result: ['Aa', 'Bb', 'Cc', 'Dd']


**Q.3. How are lambda functions typically used in Python? Provide an example use case?**

**Answer:**<br> 
In Python, an anonymous function may be a function that’s defined without a reputation. While defining normal functions, we are using the def keyword in Python, but while defining anonymous functions we are using the lambda keyword.
Syntax of Lambda Function in Python:<br>
**lambda arguments: expression**<br>
Notice, there can be any number of arguments but can contain only a single expression. There is no return statement which is usually present in the def function syntax. The function will simply return the expression value even when there is no return statement.

In [12]:
# calculate squares using lambda
squares = lambda x: x*x
print('Using lambda: ', squares(5))

Using lambda:  25


**Need for Lambda Functions**<br>
**There are at least 3 reasons:**<br>
1.Lambda functions reduce the number of lines of code when compared to normal python function defined using def keyword. But this is not exactly true because, even functions defined with def can be defined in one single line. But generally, def functions are written in more than 1 line.

2.They are generally used when a function is needed temporarily for a short period of time, often to be used inside another function such as filter, map and reduce.

3.Using lambda function, you can define a function and call it immediately at the end of definition. This can’t be done with def functions.

**When Should You Use a Lambda Function?**<br>
You should use the lambda function to create simple expressions. For example, expressions that do not include complex structures such as if-else, for-loops, and so on.<br>
So, for example, if you want to create a function with a for-loop, you should use a user-defined function.

**The following scenario is suitable for lambda expressions:**<br>
1.What you need to do is unimportant: functions don’t need a name.<br>
2.Using lambda expressions makes the code more easier to understand than the function name which you can think of.<br>
3.You’re pretty sure that there’s not a function that meets your needs.<br>
4.Everyone on your team knows lambda expressions and agrees to use it.<br>
But lambda is not recommended unless you know what you’re doing. 

**Q.4. What are the advantages and limitations of lambda functions compared to regular functions in
Python?**

**Answer:**
**Lambda Advantages:**<br>
1.The code is simple and clear.<br>
2.No additional variables are added.<br>

**Lambda Disadvantages:**<br>
1.Lambda expressions are a strange and unfamiliar syntax to many Python programmers.<br>
2.Lambda functions themselves lack names and documentation, meaning that the only way to know what they do is to read the code.<br>
3.Lambda expressions can only contain one statement, so some readable language features, such as tuple unpacking, cannot be used with them.<br>
4.Lambda functions can often be replaced with existing functions in the standard library or Python built-in functions.<br>

**Major Differences Between Lambda Expressions And Regular Functions:**<br>
1.Can be passed immediately (without variables).<br>
2.Only one line of code can be included internally.<br>
3.Automatic return of results.<br>
4.There is neither a document string nor a name.<br>

**Q.5. Are lambda functions in Python able to access variables defined outside of their own scope?
Explain with an example.**

**Answer:**

Yes, lambda functions in Python can access variables defined outside of their own scope. This is achieved through a feature called **"lexical scoping"** or **"closure"** in Python. A closure is a function object that remembers values in the enclosing scope even if they are not present in memory.<br>
Here's an example to illustrate how a lambda function can access variables from its outer scope:

In [16]:
def outer_function():
    x = 10
    lambda_func = lambda y: x + y
    return lambda_func

closure = outer_function()
result = closure(5)
print(result)

15


In this example, the **outer_function** defines a local variable **x** with a value of 10. It also defines a lambda function **lambda_func** that takes a parameter **y** and returns the sum of **x** and **y**. The lambda function is then returned from the **outer_function**.<br>

When we call **outer_function()** and assign the returned lambda function to the variable **closure**, the value of **x** (which is 10) is "captured" by the lambda function. This means that even after the **outer_function** has finished executing and **x** is no longer in its scope, the lambda function still has access to the value of **x**.<br>

Finally, when we call **closure(5)**, it adds the captured value of **x** (which is 10) to the provided argument **5**, resulting in the output of **15**.<br>

So, the lambda function can access the variable **x** defined in its outer scope, thanks to the concept of closures in Python.

**Q.6. Write a lambda function to calculate the square of a given number.**

**Answer:**

Certainly! Here's a lambda function that calculates the square of a given number:

In [17]:
square = lambda x: x ** 2

In this lambda function, the parameter x represents the number for which we want to calculate the square. The expression x ** 2 calculates the square of x.<br>
You can use this lambda function to calculate the square of a number as follows:

In [18]:
result = square(5)
print(result) 

25


**Q.7. Create a lambda function to find the maximum value in a list of integers?**

**Answer:**

In [21]:
find_max = lambda lst: max(lst)
numbers = [5, 10, 1, 45, 8, 78, 85, 15, 65, 2, 12,]
result = find_max(numbers)
print(result)  

85


In this lambda function, the parameter lst represents the list of integers. The max(lst) function is used to find the maximum value in the list. Here I call the lambda function find_max with the list of integers numbers, and it returns the maximum value in the list, which is 85.

**Q.8. Implement a lambda function to filter out all the even numbers from a list of integers?**

**Answer:**

In [22]:
# Using filter() to filter out even numbers.
def is_even(x):
    if x%2==0:
        return True
    else:
        return False

lst = [10,23,45,46,70,99,82,90,33,57]
lst1 = list(filter(is_even, lst))
print(lst1)

[10, 46, 70, 82, 90]


**Another Example:**

In [23]:
# Passing lambda function to filter() function is more elegent :
lst = [10,23,45,46,70,99,82,90,33,57]
lst1 = list(filter(lambda x: (x%2 == 0), lst))
print(lst1)

[10, 46, 70, 82, 90]


**Q.9.Write a lambda function to sort a list of strings in ascending order based on the length of each
string?**

**Answer:**

In [24]:
sort_by_length = lambda lst: sorted(lst, key=lambda x: len(x))
strings = ["apple", "banana", "pear", "orange", "kiwi"]
result = sort_by_length(strings)
print(result) 


['pear', 'kiwi', 'apple', 'banana', 'orange']


**Explanation:**
In this lambda function, the parameter lst represents the list of strings. The sorted function is used to sort the list, and the key parameter is set to a lambda function **lambda x: len(x)**. This lambda function takes each string **x** from the list and returns its length, which is then used as the basis for sorting. Here,we call the lambda function **sort_by_length** with the list of strings words, and it returns a new list with the strings sorted in ascending order based on their lengths.

**Q.10. Create a lambda function that takes two lists as input and returns a new list containing the
common elements between the two lists?**

**Answer:**

In [25]:
find_common_elements = lambda lst1, lst2: list(filter(lambda x: x in lst2, lst1))
list1 = [1, 2, 3, 4, 5]
list2 = [4, 5, 6, 7, 8]
common_elements = find_common_elements(list1, list2)
print(common_elements)  

[4, 5]


**Explanation:**
In above lambda function, the parameters **lst1 and lst2** represent the two input lists. The filter function is used to iterate over the elements of **lst1** and apply a lambda function **lambda x: x in lst2** to each element. This lambda function checks if the element **x** is present in **lst2**. If it is, the element is included in the result list.Here, We call the **lambda function find_common_elements** with the two lists **list1 and list2**, and it returns a new list containing the common elements between the two lists, which are [4, 5].

**Q.11. Write a recursive function to calculate the factorial of a given positive integer.?**

**Answer:**

In [26]:
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n - 1)
result = factorial(5)
print(result)  

120


**Explanation:**
In this recursive function, the base case is when n is equal to 0, in which case the function returns 1. For any other positive value of n, the function recursively calls itself with n - 1 and multiplies the result by n. we call the factorial function with the argument 5, and it returns the factorial of 5, which is 120.

**Q.12. Implement a recursive function to compute the nth Fibonacci number?**

**Answer:**

In [27]:
def fibonacci(n):
    if n <= 0:
        return None  # Invalid input
    elif n == 1 or n == 2:
        return 1
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)
result = fibonacci(7)
print(result)  

13


**Explanation:**
In this recursive function, the base cases are when n is less than or equal to 0 (which is considered an invalid input) or when n is 1 or 2, in which case the function returns 1. For any other positive integer n, the function recursively calls itself with n - 1 and n - 2, and returns the sum of the results of those recursive calls.we call the fibonacci function with the argument 7, and it returns the 7th Fibonacci number, which is 13.

**Q.13. Create a recursive function to find the sum of all the elements in a given list?**

**Answer:**

In [28]:
def sum_of_list_elements(lst):
    if not lst:
        return 0  # Base case: empty list
    else:
        return lst[0] + sum_of_list_elements(lst[1:])
my_list = [1, 2, 3, 4, 5]
result = sum_of_list_elements(my_list)
print(result)  

15


**Explanation:**
In this recursive function, the base case is when the input list lst is empty. In that case, the function returns 0.
For any non-empty list, the function recursively calls itself with the sublist lst[1:], which represents all elements except the first one. It then adds the first element (lst[0]) to the sum of the remaining elements computed by the recursive call. call the sum_of_list_elements function with the list [1, 2, 3, 4, 5], and it returns the sum of all the elements in the list, which is 15.

**Q.14. Write a recursive function to determine whether a given string is a palindrome?**

**Answer:**

In [29]:
def is_palindrome(string):
    if len(string) <= 1:
        return True  # Base case: empty string or single character
    elif string[0] != string[-1]:
        return False  # Base case: first and last characters don't match
    else:
        return is_palindrome(string[1:-1])
result = is_palindrome("racecar")
print(result)  

result = is_palindrome("hello")
print(result)  

True
False


**Explanation:**
In this recursive function, there are three base cases:<br>
1.If the string is empty or contains a single character, it is considered a palindrome, so the function returns True.<br>
2.If the first and last characters of the string don't match, it is not a palindrome, so the function returns False.<br>
3.For any other string, the function recursively calls itself with the substring string[1:-1], which represents the string with the first and last characters removed.<br>

In the first example, Here, I call the is_palindrome function with the string "racecar", and it returns True since "racecar" is a palindrome.

In the second example, Here, I call the function with the string "hello", and it returns False since "hello" is not a palindrome.

**Q.15.Implement a recursive function to find the greatest common divisor (GCD) of two positive integers?**

**Answer:**

In [30]:
def gcd(a, b):
    if b == 0:
        return a  # Base case: b is 0, so a is the GCD
    else:
        return gcd(b, a % b)
result = gcd(24, 36)
print(result)  

12


In [31]:
result = gcd(49, 14)
print(result)  

7


**Explanation:**
In this recursive function, the base case is when b is equal to 0. In this case, the function returns a as a is the GCD of the two numbers.

For any other positive integers a and b, the function recursively calls itself with b and a % b, which represents the remainder when a is divided by b. The GCD remains the same, but the function swaps the values of a and b in each recursive call until the base case is reached.

In the first example,Here, I call the gcd function with the two numbers 24 and 36, and it returns the GCD of the two numbers, which is 12.

In the second example, Here, I call the function with the numbers 49 and 14, and it returns the GCD of the two numbers, which is 7.